meetings/working-groups/collection-literals/CL-2023-05-26.md

```diff
@@ -0,0 +1,117 @@
+
+## Agenda
+
+* Codegen patterns and APIs for efficient construction
+* Open questions on dictionary literals
+* Cut line for C# 12
+
+## Discussion
+
+### Capacity of constructed collections
+
+If an exact length is able to be calculated for the constructed collection, the capacity could be set ahead of time to avoid resizing the backing storage of the collection. But if the constructed collection escapes outside the method body, the next .Add would trigger a resize to double the backing storage size.
+
+People will be able to write better code by hand than the compiler can generate if they have outside knowledge of the eventual state of the collection during its life past the construction of the collection literal. This would be a caveat to a statement about the collection literals feature generating optimal code. This is fine; the compiler can strike a good balance, and people can drop down to manual construction of the collection if they need more control over the initial capacity.
+
+When an exact count is known for a collection literal but the compiler can't prove that it is not potentially resized later, or when only a minimum count is known for a collection literal, the compiler could for example set the initial capacity to the best-fitting power of two. This would skip some early resizes while also amortizing resizes that could happen if the collection is modified after the collection literal is created.
+
+### Unknown-length construction
+
+When a struct-typed collection is spread, codegen could call `AddRange(IEnumerable<T>)` on the collection being constructed. This would cause the collection being spread to be boxed. To avoid boxing, codegen could loop over the spread collection and call `Add(T)` repeatedly on the collection being constructed.
+
+This would work fine when the constructed collection has a writeable Capacity property because codegen could set the capacity before calling `Add(T)` in a loop. However, if there's no writeable Capacity property, we're better off calling `AddRange` and boxing, because the `AddRange` implementation is the only way to intelligently avoid intermediate resizes. (In this situation, the thing being spread has a known length but the collection literal as a whole has an unknown length due to an earlier spread.)
+
+#### Conclusion (unknown-length construction)
+
+Stick with building using Add and AddRange. This is good enough for 95% of customers. Allow the compiler to special-case codegen for widely-used collections to improve performance.
+
+The compiler should prefer calling `AddRange(ReadOnlySpan<T>)` if such a method exists and the thing being spread is a span or exposes a span.
+
+### Known-length construction
+
+The ideal API for constructing collections with contiguous backing storage is to pass the capacity and receive a span to the backing storage. To enable creation methods to be in the same class while having the same name, the collection could be passed back using an `out` parameter.
```

Comment:
>To enable creation methods to be in the same class while having the same name [](http://example.com/codeflow?start=142&length=77)

I initially interpreted "to be in the same class" as "in the same class _as the collection being created_".

I'm not sure how to make it clearer, but perhaps: "To allow multiple creation methods to be declared in the same class with the same name, ..."